cdc: use CDC schema that is compatible with the base schema

mlitvk opened the pull request:
When generating CDC log mutations for some base mutation, use a CDC schema that is compatible with the base schema.

The compatible CDC schema has for every base column a corresponding CDC column with the same name. If using a non-compatible schema, we may encounter a situation, especially during ALTER, that we have a mutation with a base column set with some value, but the CDC schema doesn't have a column by that name. This would cause the user request to fail with an error.

We add to the schema object a schema_ptr that for CDC-enabled tables points to the schema object of the CDC table that is compatible with the schema. It is set by the schema merge algorithm when creating the schema for a table that is created or altered. We use the fact that a base table and its CDC table are created and altered in the same group0 operation, and this way we can find and set the cdc schema for a base table.

When transporting the base schema as a frozen schema between shards, we transport with it the frozen cdc schema as well.

The patch starts with a series of refactoring commits that make extending the frozen schema easier and cleans up some duplication in the code about the frozen schema. We combine the two types `frozen_schema_with_base_info` and `view_schema_and_base_info` to a single type `extended_frozen_schema` that holds a frozen schema with additional data that is not part of the schema mutations but needs to be transported with it to unfreeze it - base_info, and the frozen cdc schema which is added in a later commit.

Fixes https://github.com/scylladb/scylladb/issues/26405

backport not needed - enhancement

mlitvk commented:
> The code looks ok. What I don't quite understand is how we can write to a schema that is not yet ack:ed by all nodes? Are we no longer (?) waiting for all nodes to ack a schema update before allowing writing using said schema? (Even a node coming up after being down should methinks sync its schema:s before being allowed to accept writes, no?)

I'm not sure how it works exactly, but I was able to reproduce the crash both when adding and dropping a column:
1. [adding new column](https://github.com/scylladb/scylla-enterprise/issues/5540#issuecomment-3052515373): from my testing it seems that ALTER completes only after the schema is ready on all nodes, so if you start writing to the new column only after ALTER completes then I think it should be ok. But it's still possible to write to the new schema before ALTER completes (by a not well-behaved writer probably).
2. [dropping a column](https://github.com/scylladb/scylla-enterprise/issues/5540#issuecomment-3052639792): this one I think is more problematic, because you write to an already existing schema of the base table that has the column while it's being dropped, and the CDC transformer picks the latest CDC schema from the db that may not have the column.

elcallio commented:
Ok, but if you write to a schema before an alter is complete, you are intentionally breaking things. I don't think we should support this. We could simply just skip the column in CDC if we get a missing column, yes (instead of crash), but why bother trying to "fix" this. If you write using an altered schema before it has propagated to the replica you end up on (ugh), does the write even complete as expected? 

As for dropping, why is this a problem? CDC will simply not have any data for a column it expected to find? No?

elcallio commented:
The CDC schema is generates such that it is included in the same schema mutation set. Do we now distribute the (full) schema set non-atomically? 
In any case, if you detect that you are missing a column (or for that matter that you have a type mismatch or similar), you can either fail the write (user exception), or delay and wait for a schema update and retry, no?

piodul commented:
> The CDC schema is generates such that it is included in the same schema mutation set. Do we now distribute the (full) schema set non-atomically?

We do distribute base + CDC log mutations together. However, we do not update the in-memory structures atomically, a preemption might happen between updating the CDC log schema (which might happen first) and the base table schema. AFAIK this is what https://github.com/scylladb/scylladb/pull/24886 was supposed to fix.

In any case, the CDC transformer code fetches the CDC log schema by name, like this:

```
        , _log_schema(ctx._proxy.get_db().local().find_schema(_schema->ks_name(), log_name(_schema->cf_name())))
```

The schema returned by the lookup might already be newer than the one that would be the "proper" one for `_schema`.

> In any case, if you detect that you are missing a column (or for that matter that you have a type mismatch or similar), you can either fail the write (user exception), or delay and wait for a schema update and retry, no?

Perhaps throwing would be a good alternative for a backportable fix. Usually, when you write to a column which does not exist, you get an error that the column does not exist. If it exists in the base table but not in the CDC table then we could raise a similar error there.

piodul commented on cdc/log.cc:
```diff
@@ -1474,7 +1479,7 @@ class transformer final : public change_processor {
         , _schema(std::move(s))
         , _dk(std::move(dk))
-        , _log_schema(ctx._proxy.get_db().local().find_schema(_schema->ks_name(), log_name(_schema->cf_name())))
+        , _log_schema(local_schema_registry().get(*_schema->cdc_version()))
```
Yes, keeping the schema_ptr would probably be better. I think that's what we do in case of views, they point to their base table in this way.

mlitvk commented:
posted a simpler backportable PR to just throw an error instead of crashing: https://github.com/scylladb/scylladb/pull/24986

mlitvk commented:
* changed to store a schema_ptr to the cdc schema in the base schema, instead of schema version previously
* added some commits to refactor the frozen schema

mlitvk commented on service/migration_manager.cc:
```diff
@@ -1080,20 +1080,19 @@ static future<schema_ptr> get_schema_definition(table_schema_version v, locator:
             // That means the column mapping for the schema should always be inserted
             // with TTL (refresh TTL in case column mapping already existed prior to that).
-            auto us = s.unfreeze(db::schema_ctxt(proxy));
+            // We don't set the CDC schema here because it's not included in the RPC, but with
+            // schema on raft we shouldn't get here anyway because we do a group0 barrier which
+            // should ensure the schema is added to the registry together with the CDC schema.
+            auto us = s.unfreeze(db::schema_ctxt(proxy), nullptr);
```
I'm not sure about this
are schema pulls always disabled now? or is it possible we pull some schema, maybe during upgrade or recovery, and use it to generate cdc?

gleb-cloudius replied:
Schema pulls are always disabled if a cluster is in raft mode. Clusters that are not in raft mode will have schema pulls enabled. During upgrade schema pulls will be enabled up until the point the cluster is fully upgraded. New clusters on all supported versions will have it always disabled. Hope this covers all scenarios.

mlitvk replied:
thanks. so if I understand correctly we still need to support the cluster running not in raft mode.
in this case I think we may not always be able to set a compatible cdc schema, and if it's not set we can revert to using the current cdc schema like before.

mlitvk replied:
I updated the PR to handle the case that we run without raft and get a base schema without its cdc schema by schema pull.
in this case the pointer to the cdc schema can be null, and the cdc transformer will lookup the current cdc schema from the db.

gleb-cloudius replied:
We do not support schema changing while in recovery.

gleb-cloudius replied:
Probably the test should check if schema version actually changed. I think it will not. You are right that we need to throw an error in this case though.

avikivity commented:
How did we happen to have inconsistent schemas?

mlitvk commented:
> How did we happen to have inconsistent schemas?

because we generate the base mutations, then at some later point we generate the cdc mutations using the current cdc schema. the table could be altered in between.
original issues: https://github.com/scylladb/scylladb/issues/24952 and https://github.com/scylladb/scylla-enterprise/issues/5540

mlitvk commented:
> > > How did we happen to have inconsistent schemas?
> > 
> > 
> > because we generate the base mutations, then at some later point we generate the cdc mutations using the current cdc schema. the table could be altered in between. original issues: #24952 and [scylladb/scylla-enterprise#5540](https://github.com/scylladb/scylla-enterprise/issues/5540)
> 
> I see. Can't we capture both schemas at the same time?

I think we can. I'm thinking we can introduce `cdc_augmentable_mutation` which is a mutation and its cdc schema ptr, and pass it all the way from when we create the mutation through the different flows where we augment it with cdc mutations.
I'm not sure if it's better

avikivity commented:
> > > > How did we happen to have inconsistent schemas?
> > > 
> > > 
> > > because we generate the base mutations, then at some later point we generate the cdc mutations using the current cdc schema. the table could be altered in between. original issues: #24952 and [scylladb/scylla-enterprise#5540](https://github.com/scylladb/scylla-enterprise/issues/5540)
> > 
> > 
> > I see. Can't we capture both schemas at the same time?
> 
> I think we can. I'm thinking we can introduce `cdc_augmentable_mutation` which is a mutation and its cdc schema ptr, and pass it all the way from when we create the mutation through the different flows where we augment it with cdc mutations. I'm not sure if it's better

Seems much simpler, no?

avikivity commented:
> > > > > > > How did we happen to have inconsistent schemas?
> > > > > > 
> > > > > > 
> > > > > > because we generate the base mutations, then at some later point we generate the cdc mutations using the current cdc schema. the table could be altered in between. original issues: #24952 and [scylladb/scylla-enterprise#5540](https://github.com/scylladb/scylla-enterprise/issues/5540)
> > > > > 
> > > > > 
> > > > > I see. Can't we capture both schemas at the same time?
> > > > 
> > > > 
> > > > I think we can. I'm thinking we can introduce `cdc_augmentable_mutation` which is a mutation and its cdc schema ptr, and pass it all the way from when we create the mutation through the different flows where we augment it with cdc mutations. I'm not sure if it's better
> > > 
> > > 
> > > Seems much simpler, no?
> 
> it touches a lot of things, starting from all places that generate mutations - cql statements, alternator executor, internal mutations - through different flows until the cdc augmentation - mutate, mutate with triggers, mutate atomically, paxos

Yes, but the end result?

> 
> > Or even: capture the CDC schema in the base schema. Not sure if it will cause a cyclic dependency.
> 
> This is basically what this PR is doing, it captures the CDC schema in the base schema, but at the time the schemas are created.

Then I'll check it out, I understood something else.

The cover letter gives a different impression.

avikivity commented on db/schema_applier.cc:
```diff
@@ -579,6 +581,32 @@ in_progress_types_storage_per_shard& in_progress_types_storage::local() {
+static auto extract_cdc(
+        std::map<table_id, schema_mutations>& tables_before,
+        std::map<table_id, schema_mutations>& tables_after)
+{
```
It's better for caller readability to make this a pure function that returns four maps.

Also better to return a struct instead of auto.

mlitvk replied:
changed to a pure function that returns a struct

lsfreitas commented:
CI failed on this PR due to an issue introduced and already reverted causing a wrong status. Please re-trigger CI

piodul commented:
@wmitros please take a look and review, you worked on the issue of inconsistent base/view schema pointers, this one is about CDC but the issue and some parts of the code touched are similar

wmitros reviewed (commented):
I posted some questions but looks good. 
The fact that the new `_cdc_schema` field in `schema` can sometimes be `nullptr` even if the table uses cdc is a bit concerning. We had issues with the base_info field not being set which might be similar, but it's less likely to be a problem here - we always update the base and cdc schemas in one group0 mutation and nowadays the schema mutations are applied atomically. Also, eventually, the `_cdc_schema` is only used for constructing the cdc `transformer` so it should be easier to identify where we can skip setting the `_cdc_schema`, as we did in the patch.
+nice cleanup with the extended_frozen_schema

mlitvk commented on db/schema_tables.cc:
```diff
@@ -2096,7 +2097,9 @@ future<schema_ptr> create_table_from_name(distributed<service::storage_proxy>& p
-    co_return create_table_from_mutations(ctxt, std::move(sm), ctxt.user_types(), nullptr/*TODO cdc_schema*/);
+    // The CDC schema is set to nullptr because we don't have it yet, but we will
+    // check and update it soon if needed in create_tables_from_tables_partition.
+    co_return create_table_from_mutations(ctxt, std::move(sm), ctxt.user_types(), nullptr);
```
`create_table_from_name` is only called from `create_table_from_table_row`
`create_table_from_table_row` is only called from `create_tables_from_tables_partition`
the schema created here is returned to `create_tables_from_tables_partition` where we collect all the tables schemas and extend them with cdc schemas

mlitvk commented on db/schema_applier.cc:
```diff
@@ -634,11 +634,46 @@ static future<> merge_tables_and_views(distributed<service::storage_proxy>& prox
     auto& local_tables = diff.tables_and_views.local().tables;
     auto& local_cdc = diff.tables_and_views.local().cdc;
 
+    // Create CDC tables before non-CDC base tables, because we want the base tables with CDC enabled
```
fixed, thanks

mlitvk commented on db/schema_tables.cc:
```diff
@@ -2114,10 +2117,26 @@ constexpr size_t max_concurrent = 8;
+        if (auto it = tables.find(cdc::log_name(name)); it != tables.end()) {
+            cdc_schema = it->second;
+        }
+        schema_ptr extended_cfm = cdc_schema ? cfm->make_with_cdc(cdc_schema) : cfm;
```
if it's not found then the cdc transformer falls back to the old behavior, which is to get the current cdc schema from db

mlitvk commented on test/cluster/test_cdc_with_alter.py:
```diff
@@ -73,3 +73,39 @@ def do_writes():
+            f"Unexpected values in column 'a' of CDC log: {[row.a for row in cdc_rows]}"
+
+        matches = await log.grep("has no CDC schema set")
+        assert len(matches) == 0, "Found unexpected log messages indicating missing CDC schema"
```
fixed

denesb commented:
I would queue but I cannot determine what is the state of this PR. Half-a-dozen people participated in the discussion that spawns 70+ comments and there is no explicit approval.

mlitvk commented:
> @mykaul I see that this PR was marked for 2025.4 - do we really want to backport this now? The bug that this PR fixes is not a regression so I think it would benefit from some soak time on master.
> 
> @mlitvk please correct me if I'm wrong, but the current impact of the bug is that some write requests might fail while dropping a column on the base table - and that's it, right? Nothing more serious will happen like nodes crashing etc.? If so, then it's not a common case and the impact is not high, so really not sure that we should backport.

yes, I agree it's better not to backport it. the issue is not severe, it's just a write failure in a not very common scenario, and the change is not trivial

piodul commented:
Queued

mlitvk commented on cdc/log.cc:
```diff
@@ -1519,7 +1519,7 @@ class transformer final : public change_processor {
         , _schema(std::move(s))
         , _dk(std::move(dk))
-        , _log_schema(ctx._proxy.get_db().local().find_schema(_schema->ks_name(), log_name(_schema->cf_name())))
+        , _log_schema(_schema->cdc_schema() ? _schema->cdc_schema() : ctx._proxy.get_db().local().find_schema(_schema->ks_name(), log_name(_schema->cf_name())))
```
opened an issue about removing the fallback: https://github.com/scylladb/scylladb/issues/27012
